File: packages/autonomo-ruby/lib/autonomo/actions.rb
# frozen_string_literal: true

module Autonomo
  # Result of a custom action
  class ActionResult
    attr_reader :success, :message, :error, :data

    def initialize(success:, message: nil, error: nil, data: nil)
      @success = success
      @message = message
      @error = error
      @data = data
    end

    def to_h
      result = { success: @success }
      result[:message] = @message if @message
      result[:error] = @error if @error
      result[:data] = @data if @data
      result
    end

    def self.ok(message = nil, data: nil)
      new(success: true, message: message, data: data)
    end

    def self.fail(error)
      new(success: false, error: error)
    end
  end

  # Metadata for a custom action - helps AI understand what it does
  class CustomActionMeta
    attr_reader :description, :args, :example

    def initialize(description: nil, args: nil, example: nil)
      @description = description
      @args = args
      @example = example
    end
  end

  # Rich custom action info returned in state
  class CustomActionInfo
    attr_reader :name, :description, :args, :example

    def initialize(name:, description: nil, args: nil, example: nil)
      @name = name
      @description = description
      @args = args
      @example = example
    end

    def to_h
      result = { name: @name }
      result[:description] = @description if @description
      result[:args] = @args if @args
      result[:example] = @example if @example
      result
    end
  end

  # Internal registered action
  class RegisteredAction
    attr_reader :handler, :meta

    def initialize(handler:, meta: nil)
      @handler = handler
      @meta = meta
    end
  end

  # Singleton registry for custom actions
  class CustomActionsRegistry
    @instance = nil

    def self.instance
      @instance ||= new
    end

    def initialize
      @actions = {}
      @listeners = []
      @mutex = Monitor.new  # Use Monitor for reentrant locking
    end

    # Register a custom action
    def register(name, meta: nil, &handler)
      @mutex.synchronize do
        @actions[name] = RegisteredAction.new(handler: handler, meta: meta)
        notify_change
      end
      -> { unregister(name) }
    end

    # Unregister a custom action
    def unregister(name)
      @mutex.synchronize do
        if @actions.delete(name)
          notify_change
        end
      end
    end

    # Execute a custom action
    def execute(name, value = nil)
      action = @mutex.synchronize { @actions[name] }
      
      unless action
        available = list.empty? ? 'none' : list.join(', ')
        return ActionResult.fail("Unknown custom action: #{name}. Available: #{available}")
      end

      begin
        action.handler.call(value)
      rescue => e
        ActionResult.fail(e.message)
      end
    end

    # Check if action exists
    def has?(name)
      @mutex.synchronize { @actions.key?(name) }
    end

    # List all action names
    def list
      @mutex.synchronize { @actions.keys }
    end

    # Get rich info about all actions (for AI discoverability)
    def get_all
      @mutex.synchronize do
        @actions.map do |name, action|
          CustomActionInfo.new(
            name: name,
            description: action.meta&.description,
            args: action.meta&.args,
            example: action.meta&.example
          )
        end
      end
    end

    # Subscribe to changes
    def on_change(&listener)
      @mutex.synchronize { @listeners << listener }
      -> { @mutex.synchronize { @listeners.delete(listener) } }
    end

    private

    def notify_change
      @listeners.each(&:call)
    end
  end

  # Module-level helpers
  class << self
    def custom_actions
      CustomActionsRegistry.instance
    end

    # Register a custom action
    def register_custom_action(name, meta: nil, &handler)
      custom_actions.register(name, meta: meta, &handler)
    end
  end
end

File: packages/autonomo-swift/Sources/Autonomo/Actions.swift
import Foundation

// MARK: - Action Result

/// Result of a custom action
public struct ActionResult {
    public let success: Bool
    public let message: String?
    public let error: String?
    public let data: Any?
    
    public init(success: Bool, message: String? = nil, error: String? = nil, data: Any? = nil) {
        self.success = success
        self.message = message
        self.error = error
        self.data = data
    }
    
    public static func ok(_ message: String? = nil, data: Any? = nil) -> ActionResult {
        ActionResult(success: true, message: message, data: data)
    }
    
    public static func fail(_ error: String) -> ActionResult {
        ActionResult(success: false, error: error)
    }
    
    public func toDictionary() -> [String: Any] {
        var result: [String: Any] = ["success": success]
        if let message = message { result["message"] = message }
        if let error = error { result["error"] = error }
        if let data = data { result["data"] = data }
        return result
    }
}

// MARK: - Custom Action Metadata

/// Metadata for a custom action - helps AI understand what it does
public struct CustomActionMeta {
    public let description: String?
    public let args: [String: String]?
    public let example: [String: String]?
    
    public init(description: String? = nil, args: [String: String]? = nil, example: [String: String]? = nil) {
        self.description = description
        self.args = args
        self.example = example
    }
}

/// Rich custom action info returned in state
public struct CustomActionInfo {
    public let name: String
    public let description: String?
    public let args: [String: String]?
    public let example: [String: String]?
    
    public func toDictionary() -> [String: Any] {
        var result: [String: Any] = ["name": name]
        if let description = description { result["description"] = description }
        if let args = args { result["args"] = args }
        if let example = example { result["example"] = example }
        return result
    }
}

// MARK: - Custom Action Handler

public typealias CustomActionHandler = (String?) -> ActionResult

// MARK: - Registered Action

private struct RegisteredAction {
    let handler: CustomActionHandler
    let meta: CustomActionMeta?
}

// MARK: - Custom Actions Registry

/// Singleton registry for custom actions
public class CustomActionsRegistry {
    public static let shared = CustomActionsRegistry()
    
    private var actions: [String: RegisteredAction] = [:]
    private var listeners: [() -> Void] = []
    private let lock = NSLock()
    
    private init() {}
    
    /// Register a custom action
    @discardableResult
    public func register(_ name: String, handler: @escaping CustomActionHandler, meta: CustomActionMeta? = nil) -> () -> Void {
        lock.lock()
        actions[name] = RegisteredAction(handler: handler, meta: meta)
        lock.unlock()
        notifyChange()
        return { [weak self] in self?.unregister(name) }
    }
    
    /// Unregister a custom action
    public func unregister(_ name: String) {
        lock.lock()
        let removed = actions.removeValue(forKey: name) != nil
        lock.unlock()
        if removed { notifyChange() }
    }
    
    /// Clear all registered custom actions
    public func clear() {
        lock.lock()
        actions.removeAll()
        lock.unlock()
        notifyChange()
    }
    
    /// Execute a custom action
    public func execute(_ name: String, value: String? = nil) -> ActionResult {
        lock.lock()
        let action = actions[name]
        lock.unlock()
        
        guard let action = action else {
            let available = list().isEmpty ? "none" : list().joined(separator: ", ")
            return .fail("Unknown custom action: \(name). Available: \(available)")
        }
        
        do {
            return action.handler(value)
        } catch {
            return .fail(error.localizedDescription)
        }
    }
    
    /// Check if action exists
    public func has(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return actions[name] != nil
    }
    
    /// List all action names
    public func list() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return Array(actions.keys)
    }
    
    /// Get rich info about all actions (for AI discoverability)
    public func getAll() -> [CustomActionInfo] {
        lock.lock()
        defer { lock.unlock() }
        return actions.map { name, action in
            CustomActionInfo(
                name: name,
                description: action.meta?.description,
                args: action.meta?.args,
                example: action.meta?.example
            )
        }
    }
    
    /// Subscribe to changes
    @discardableResult
    public func onChange(_ listener: @escaping () -> Void) -> () -> Void {
        lock.lock()
        listeners.append(listener)
        let index = listeners.count - 1
        lock.unlock()
        return { [weak self] in
            self?.lock.lock()
            if let self = self, index < self.listeners.count {
                self.listeners.remove(at: index)
            }
            self?.lock.unlock()
        }
    }
    
    private func notifyChange() {
        lock.lock()
        let currentListeners = listeners
        lock.unlock()
        for listener in currentListeners {
            listener()
        }
    }
}

// MARK: - Helper

/// Register a custom action
@discardableResult
public func registerCustomAction(_ name: String, handler: @escaping CustomActionHandler, meta: CustomActionMeta? = nil) -> () -> Void {
    CustomActionsRegistry.shared.register(name, handler: handler, meta: meta)
}
